clis/teliod/src/core_api.rs

```diff
@@ -0,0 +1,343 @@
+) -> Result<DeviceIdentity, Error> {
+    let mut identity_path = dirs::data_local_dir().ok_or(Error::NoDataLocalDir)?;
+    identity_path.push("teliod");
+    if !identity_path.exists() {
+        let _ = create_dir_all(&identity_path);
+    }
+    identity_path.push(&format!("{interface_name}.json"));
+
+    let mut device_identity = DeviceIdentity::new(&identity_path).await;
+
+    if device_identity.machine_identifier.is_none() {
+        device_identity.machine_identifier = Some(
+            match fetch_identifier_with_exp_backoff(
+                auth_token,
+                device_identity.private_key.public(),
+            )
+            .await
+            {
+                Ok(id) => id,
+                Err(e) => {
+                    debug!("Unable to load identifier due to {e}. Registering ...");
+                    register_machine_with_exp_backoff(
+                        &device_identity.hw_identifier,
+                        device_identity.private_key.public(),
+                        auth_token,
+                    )
+                    .await?
+                }
+            },
+        );
+    }
+
+    let mut backoff = build_exp_backoff()?;
+    let mut retries = 0;
+    loop {
+        let status = update_machine(&device_identity, auth_token).await?;
+        match status {
+            StatusCode::OK => break,
+            StatusCode::FORBIDDEN | StatusCode::UNAUTHORIZED | StatusCode::BAD_REQUEST => {
+                // exit daemon
+                return Err(Error::UpdateMachineError(status));
+            }
+            StatusCode::NOT_FOUND => {
+                debug!("Unable to update. Registering machine ...");
+                // Retry machine update after registering. To make sure everything was successful
+                // If registering fails. Close the daemon
+                device_identity.machine_identifier = Some(
+                    register_machine_with_exp_backoff(
+                        &device_identity.hw_identifier,
+                        device_identity.private_key.public(),
+                        auth_token,
+                    )
+                    .await?,
+                );
+            }
+            _ => {
+                // Retry with exp back-off
+                let _ =
+                    retry_after_backoff_delay(&mut backoff, &mut retries, "update machine").await?;
+            }
+        }
+    }
+
+    let mut options = OpenOptions::new();
+    // User has read/write but others have none
+    options.mode(0o600);
+    let mut file = options.create(true).write(true).open(identity_path)?;
+    serde_json::to_writer(&mut file, &device_identity)?;
+    Ok(device_identity)
+}
+
+async fn fetch_identifier_with_exp_backoff(
+    auth_token: &str,
+    public_key: PublicKey,
+) -> Result<MachineIdentifier, Error> {
+    let mut retries = 0;
+    let mut backoff = build_exp_backoff()?;
+    loop {
+        match fetch_identifier_from_api(auth_token, public_key).await {
```

Comment:
Maybe same here? Not much sense in retrying if we get `Invalid authorization header` or `Invalid credentials`

Reply:
Done!